### PR TITLE
[SPARK-48817][SQL] Eagerly execute union multi commands together

### PR DESCRIPTION
**Pull request by wForget**
### What changes were proposed in this pull request?

Eagerly execute union multi commands together.

### Why are the changes needed?
MultiInsert is split to multiple sql executions, resulting in no exchange reuse.

Reproduce sql:

```
create table wangzhen_t1(c1 int);
create table wangzhen_t2(c1 int);
create table wangzhen_t3(c1 int);
insert into wangzhen_t1 values (1), (2), (3);

from (select /*+ REPARTITION(3) */ c1 from wangzhen_t1)
insert overwrite table wangzhen_t2 select c1
insert overwrite table wangzhen_t3 select c1; 
```

In Spark 3.1, there is only one SQL execution and there is a reuse exchange.

image

However, in Spark 3.5, it was split to multiple executions and there was no ReuseExchange.

image
image

### Does this PR introduce _any_ user-facing change?

yes,  multi  inserts will executed in one execution.

### How was this patch tested?

added unit test

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by wForget** [comment]
It seems to be caused by https://github.com/apache/spark/pull/32513

**Comment by wForget** [comment]
@cloud-fan @beliefer Could you please take a look?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 1]
```diff
@@ -129,7 +129,15 @@ class QueryExecution(
...
         result.toImmutableArraySeq)
-    case other => other
+    }
+    p transformDown {
+      case u @ Union(children, _, _) if children.forall(_.isInstanceOf[Command]) =>
+        eagerlyExecute(u, "commands", CommandExecutionMode.SKIP)
+      case c: Command =>
+        val name = commandExecutionName(c)
+        eagerlyExecute(c, name, CommandExecutionMode.NON_ROOT)
+      case other => other
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:139]: nit: unnecessary default pattern

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 2]
```diff
@@ -110,15 +110,15 @@ class QueryExecution(
     case _ => "command"
   }
 
-  private def eagerlyExecuteCommands(p: LogicalPlan) = p transformDown {
-    case c: Command =>
+  private def eagerlyExecuteCommands(p: LogicalPlan) = {
+    def eagerlyExecute(p: LogicalPlan,
+        name: String, mode: CommandExecutionMode.Value): LogicalPlan = {
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:115]:
```scala
def eagerlyExecute(
  p: LogicalPlan,
  name: String,
  mode: CommandExecutionMode.Value)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 3]
```diff
@@ -129,7 +129,15 @@ class QueryExecution(
         qe.commandExecuted,
         qe.executedPlan,
         result.toImmutableArraySeq)
-    case other => other
+    }
+    p transformDown {
+      case u @ Union(children, _, _) if children.forall(_.isInstanceOf[Command]) =>
+        eagerlyExecute(u, "commands", CommandExecutionMode.SKIP)
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:135]: `commands` -> `multi-commands`

**Review by ulysses-you — COMMENTED** [review]
lgtm except some minor comments

**Review by ulysses-you — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merged to master

**Comment by cloud-fan** [comment]
late LGTM

**Pull request closed without merging**